auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -1,25 +1,217 @@
     @Override
-    public Principal authenticate(final ICredential credential) {
-        throw new IllegalStateException("Not implemented");
+    public Principal authenticate(final ICredential credential) throws AuthenticationException {
+        OIDCUserPrincipalImpl user;
+        try {
+            userIdentity = getIdentityInformation(credential);
+            user = (OIDCUserPrincipalImpl) OIDCUserPrincipalImpl
+                .newUser()
+                .withAuthenticator(this)
+                .withUsername(parseUsername())
+                .withCredentials(credential)
+                .withEmail(parseEmail())
+                .withRoles(parseRoles())
+                .build();
+        } catch (JSONException e) {
+            throw new AuthenticationException(e.getMessage(), e.getCause());
+        }
+    return user;
+    }
+
+    /**
+     * Parses the user's username from the user identity {@link #userIdentity}
+     *
+     * @return user's username
+     * @throws JSONException
+     */
+    private String parseUsername() throws JSONException {
+        String username = "Unknown Username";
+        if (userIdentity != null) {
+            // get the users username
+            if (userIdentity.has(USERNAME) && userIdentity.getString(USERNAME).length() > 0) {
+                username = userIdentity.getString(USERNAME);
+            }
+        }
+        return username;
+    }
+
+    /**
+     * Parses the user's email address from the user identity {@link #userIdentity}
+     *
+     * @return user's email address
+     * @throws JSONException
+     */
+    private String parseEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    /**
+     * Parses the user's roles from the user identity {@link #userIdentity}
+     *
+     * @return user's roles
+     * @throws JSONException
+     */
+    private Collection<IRole> parseRoles() throws JSONException {
+        Collection<IRole> roles = new ArrayList<IRole>();
+        if (userIdentity != null) {
+            Map<RoleKey, IRole> realmRoles = parseRealmRoles();
+            if (realmRoles != null) {
+                roles.addAll(realmRoles.values());
+            }
+            Map<RoleKey, IRole> clientRoles = parseClientRoles();
+            if (clientRoles != null) {
+                roles.addAll(clientRoles.values());
+            }
+        }
+        return roles;
+    }
+
+    /**
+     * Parses the user's realm roles from the user identity {@link #userIdentity}
+     *
+     * @return user's realm roles
+     * @throws JSONException
+     * @see RoleKey
+     * @see RealmRole
+     */
+    private Map<RoleKey, IRole> parseRealmRoles() throws JSONException {
```

Comment:
not sure why we return a Map here. Why not just return an array or list of RealmRole?